[mono] Implement `AssemblyExtensions.TryGetRawMetadata`.

teo-tsirpanis opened the pull request.

teo-tsirpanis committed:
[mono] Implement `AssemblyExtensions.TryGetRawMetadata`.

jkoritzinsky commented on src/mono/mono/metadata/icall.c:
```diff
@@ -4743,6 +4743,22 @@ ves_icall_System_Reflection_Assembly_InternalGetReferencedAssemblies (MonoReflec
 }
 
+MonoBoolean
+ves_icall_System_Reflection_RuntimeAssembly_InternalTryGetRawMetadata (MonoQCallAssemblyHandle assembly_h, gpointer_ref blob, gint32_ref length, MonoError *error)
+{
+	MonoAssembly *assembly = assembly_h.assembly;
+	MonoImage *image = assembly->image;
+
+	if (image_is_dynamic (image)) {
+		return FALSE;
+	}
+
+	*blob = image->raw_metadata;
+	*((guint32*)length) = image->raw_metadata_len;
+
+	return TRUE;
+}
```
Are you sure that this handles the hot reload case? I think this would give back the metadata for the unedited image with Mono's hot reload implementation.

teo-tsirpanis replied:
I have no idea. What does CoreCLR do in hot-reloaded assemblies?

jkoritzinsky replied:
For CoreCLR, we have an editable metadata system, so RefEmit and Hot Reload go down the same path (no metadata blob available to pass up to callers).

For Mono, their RefEmit solution is a completely separate lookup system, so Hot Reload uses a separate mechanism that works via "layers".

jkoritzinsky replied:
So for Mono, you need to check the `has_updates` bit on the MonoImage. If that's true, this function needs to return false.

teo-tsirpanis replied:
Done.

jkotas replied:
> For CoreCLR, we have an editable metadata system, so RefEmit and Hot Reload go down the same path (no metadata blob available to pass up to callers).

If I am reading the correctly, CoreCLR returns the original metadata blob for hot reload assemblies from this API.

> you need to check the has_updates bit on the MonoImage.

I think it is unnecessary difference from the CoreCLR behavior. I do not see a block like this in the CoreCLR implementation.

teo-tsirpanis committed:
Do not support getting raw metadata of hot-reloaded assemblies.
And return if the metadata blob pointer is not null, matching CoreCLR implementation.

teo-tsirpanis committed:
Remove unnecessary checks.

lewing committed:
Merge branch 'main' into mono-raw-md

teo-tsirpanis committed:
Merge branch 'main' into mono-raw-md

jkotas committed:
Merge branch 'main' into mono-raw-md
